File: Java/BaekJoon/Silver/S2_16401.java
package BaekJoon.Silver;

import java.util.Arrays;
import java.util.Scanner;

public class S2_16401 {
    public static int m, n;
    public static int[] snack = new int[1000000];

    public static boolean check(int leng) {
        int cnt = 0;

        for (int i = 0; i < n; i++) {
            if (snack[i] >= leng) {
                cnt += snack[i] / leng;
            }
        }

        return cnt >= m;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        m = sc.nextInt();
        n = sc.nextInt();

        for (int i = 0; i < n; i++) {
            int temp = sc.nextInt();
            snack[i] = temp;
        }

        Arrays.sort(snack, 0, n);

        int lo = 1;
        int hi = snack[n - 1];

        int leng = 0;

        while(lo <= hi) {
            int mid = (lo + hi) / 2;

            if (check(mid)) {
                leng = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        System.out.println(leng);

        sc.close();
    }
}

File: C/BaekJoon/Silver/S2_16401.c
#include <stdio.h>
#include <stdlib.h>

int m, n;
int snack[1000000];

int cmp(const void* a, const void* b) {
  int x = *(int*)a;
  int y = *(int*)b;

  return x - y;
}

int check(int leng) {
  int cnt = 0;

  for (int i = 0; i < n; i++) {
    if (snack[i] >= leng) {
      cnt += snack[i] / leng;
    }
  }

  if (cnt >= m) {
    return 1;
  } else {
    return 0;
  }
}

int main() {
  scanf("%d %d", &m, &n);

  for (int i = 0; i < n; i++) {
    scanf("%d", &snack[i]);
  }

  qsort(snack, n, sizeof(int), cmp);

  int lo = 1;
  int hi = snack[n - 1];

  int leng = 0;

  while (lo <= hi) {
    int mid = (lo + hi) / 2;

    if (check(mid)) {
      lo = mid + 1;
      leng = mid;
    } else {
      hi = mid - 1;
    }
  }

  printf("%d", leng);

  return 0;
}